### DISCUSSION TITLE
Baking animation into blender file

### DISCUSSION THREAD

**Discussion by christiantownsend** [coldtype/coldtype#98] [Q&A] [answered]
Hi!

I've been attempting to get coldtype to work with blender in a way where I can export the animation (as an alembic or fbx or similar) and bring it into a different blend file with the animation baked in. I noticed in one of the examples there's an option to set `bake=True` on the animation, but when that option is enabled I always get this error:

```
>>> resave: animations\test.py
------------------------------
Traceback (most recent call last):
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\blender\watch.py", line 230, in reimport
    animation_found = self.render_current_frame(statics=True)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\blender\watch.py", line 188, in render_current_frame
    walk_to_b3d(r.baked_frames(), renderable=r)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\blender\__init__.py", line 372, in baked_frames
    return to_bake.walk(bakewalk)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\runon\runon.py", line 307, in walk
    el.walk(callback, depth=depth+1, visible_only=visible_only,
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\runon\runon.py", line 307, in walk
    el.walk(callback, depth=depth+1, visible_only=visible_only,
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\runon\runon.py", line 320, in walk
    res = callback(self, 0, dict(
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\blender\__init__.py", line 361, in bakewalk
    (p.ch(b3d(f"CTBakedAnimation_{self.name}",
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\runon\runon.py", line 832, in chain
    res = fn(self, *args)
  File "C:\Program Files\Blender Foundation\Blender 2.93\2.93\python\lib\site-packages\coldtype\blender\__init__.py", line 103, in annotate
    pen_mod(pen)
TypeError: 'str' object is not callable
```

For reference, I'm on windows and I've tried with blender 2.93 (which ships with python 3.9) and blender 3.1 (which ships with python 3.10) just to make sure that wasn't the issue since I saw using 3.9 fixed a previous error. It could also be that setting bake to True doesn't even accomplish what I'm trying to do! I'm going to try on mac next and see if I have the same issue, but I'd love to know if what I'm trying to accomplish is possible currently.

Also, my python file looks like this (which I just pulled from the examples to see if I could get a basic case working):

``` python
from coldtype import *
from coldtype.blender import *

@b3d_animation(timeline=60, bake=True) # switch to True
def baketest2(f):
    return (StSt("TEST", Font.MutatorSans(), 200,
        wdth=f.e("eeio", 1),
        wght=f.e("eeio", 2),
        ro=1)
        .align(f.a.r, tv=1)
        .pmap(lambda p: p
            .ch(b3d(lambda bp: bp
                .extrude(f.e("seio", 1, rng=(0.1, 1.5)))
                ))))
```

Thanks a ton!

**Comment by stenson** [comment, marked as answer]
Ah! Yeah that's just a good old fashioned bug in the Coldtype code — should be fixed now in the new 0.9.5 release that just went up (you can grab that with a `python -m pip install "coldtype[viewer] --upgrade` in your venv & a `b3d_python -m pip install "coldtype[viewer] --upgrade` if you have that alias set up)

Made a couple small tweaks to your code above, just because the `examples/blender/bake.py` file hadn't been upgraded in a while. The `def setup` isn't completely necessary but should handle cleaning up the state a little more thoroughly.

```python
from coldtype import *
from coldtype.blender import *

@b3d_runnable()
def setup(bw:BpyWorld):
    bw.deletePrevious("CTBakedAnimation_baketest2")

@b3d_animation(timeline=60
    , reset_to_zero=1
    , bake=True) # switch to True
def baketest2(f):
    return (StSt("TEST", Font.MutatorSans(), 200,
        wdth=f.e("eeio", 1),
        wght=f.e("eeio", 2),
        ro=1)
        .align(f.a.r, tv=1)
        .pen()
        .ch(b3d(lambda bp: bp
            .extrude(f.e("seio", 1, rng=(0.1, 1.5))))))
```

I also swapped the `.pmap` for a `.pen()` call here to turn the whole word into a single vector/object before it gets baked, just so there'd be fewer keyframes objects in the Blender file itself. That's not a necessary change either, just a thought for keeping the number of objects down.

Thanks for the question, and sorry you had to fight a bug there!

Feel free to follow-up as well if the whole bake-to-keyframes process isn't what you're expecting — I have to admit that was kind of a speculative addition to the library (hence the bugginess), since I usually don't bake anything to keyframes myself, so I'm sure there's room for improvement.

**Reply by christiantownsend** [reply]
Thanks for the quick response, I've got it baking now!

I'm definitely gonna have to do some tinkering to see if I can get it exported as a mesh sequence or something that's easy to just plop into another file from here, but I don't think that will be too much of a hassle. I'll definitely follow up if I make any discoveries on that front, but I'm imagining you're right that this isn't a super common workflow.

I've been having a blast with coldtype by the way, really appreciate you putting this out there! It's surprising how difficult it has been to work with variable type in drawbot and other preexisting tools, so this has been a super handy addition to the toolkit!
